metals/src/main/scala/scala/meta/internal/metals/ExcludedPackagesHandler.scala

```diff
@@ -0,0 +1,71 @@
+package scala.meta.internal.metals
+
+/**
+ * This class handles packages that should be excluded from indexing causing
+ * them to not be recommended for completions, symbol searches, and code actions.
+ *
+ * @param pkgsToExclude any exclusions that are in the `UserConfiguration` to start with.
+ */
+class ExcludedPackagesHandler(pkgsToExclude: Option[List[String]]) {
+  val defaultExclusions: List[String] = List(
+    "META-INF/", "images/", "toolbarButtonGraphics/", "jdk/", "sun/", "javax/",
+    "oracle/", "java/awt/desktop/", "org/jcp/", "org/omg/", "org/graalvm/",
+    "com/oracle/", "com/sun/", "com/apple/", "apple/"
+  )
+
+  /**
+   * Cached exclusions to make sure that we only have to process the list
+   * once. This is only updated if the user changes the exclusions in the
+   * UserConfiguration which is detected in the `workspace/didChangeConfiguration`.
+   */
+  private var cachedExclusions: List[String] = List.empty
+
+  /**
+   * More than likely a user will give us exclusions in the following format:
+   *
+   * - `akka.actor.typed.javadsl`
+   *
+   * This will make sure that they are always formatted in the way we need them
+   * to be when we are actually comparing them as paths:
+   *
+   * - `akka/actor/typed/javadsl/`
+   *
+   * @param pkgs pkgs to either add to or check against the defaults.
+   */
+  private def prepareCache(pkgs: List[String]): List[String] = {
+    pkgs.foldLeft(defaultExclusions) { (exclusions, pkg) =>
+      val replaced = pkg.replace(".", "/")
+      val formatted =
+        if (replaced.endsWith("/")) replaced
+        else replaced + "/"
+
+      if (formatted.startsWith("--")) {
+        exclusions.filterNot(_ == formatted.drop(2))
+      } else {
+        formatted :: exclusions
+      }
+    }
+  }
+
+  def update(pkgs: List[String]): Unit = cachedExclusions = prepareCache(pkgs)
+
+  /**
+   * Should the given package be excluded from indexing
+   *
+   * @param pkg package to check against excluded list
+   */
+  def isExcludedPackage(pkg: String): Boolean = {
+    if (cachedExclusions.nonEmpty) {
```

Comment:
If we did:
`def isExcludedPackage(pkg: String): Boolean = cachedExclusions.exists(excluded => pkg.startsWith(excluded))`

It would not work only in 1 case where we created this class but did not call update yet, because `cachedExclusions` would be empty list.

But if we did:
`private var cachedExclusions: List[String] = prepareCache(pkgsToExclude)`

Then this should work? 

Fine as it now, just curious what you think?

Reply:
So my main reason for doing it this is that since this is created right after the user configuration, it will get the default user configuration value, which is `None`. Then if the user has any value set, it will run the update right away as soon as the initialize happens. My thinking was then that we'd run this twice no matter what whenever it was used. So my attempt to avoid that was to hold off on running the `prepareCache` until it's needed to index the first time, or when the `initialize` process happens and the change is picked up during the `didChangeConfiguration`. So I don't think it matters a ton, but again, was trying to avoid running `prepareCache` twice right in a row. I'm open to maybe changing this, but also don't see a downside of having it this way. WDYT?

Reply:
Ahh yes right!, I did not notice it will be called twice in that case, all good :+1: